kernel/src/process/signal/pauser.rs

```diff
@@ -259,7 +255,185 @@ impl SigQueueObserver {
+///
+/// ```no_run
+/// let waker = paused_thread.waker();
+/// waker.wake_up();
+/// ```
+pub struct Pauser {
+    current_thread: Arc<Thread>,
+    waiter: Waiter,
+    sig_mask: SigMask,
+}
+
+impl !Sync for Pauser {}
+impl !Send for Pauser {}
+
+impl Pauser {
+    /// Creates a new `Pauser`.
+    ///
+    /// The `Pauser` can be interrupted by all signals
+    /// except that are blocked by current thread.
+    pub fn new() -> Self {
+        let current_thread = Thread::current().unwrap();
+        let waiter = current_thread.waker().try_create_waiter().unwrap();
+        Self {
+            current_thread: current_thread.clone(),
+            waiter,
+            sig_mask: SigMask::new_empty(),
+        }
+    }
+
+    /// Creates a new `Pauser` with current thread inputted.
+    ///
+    /// This is a more faster method compared to [`Self::new`].
+    ///
+    /// # Panics
+    /// if the inputted thread is not the current thread, this method will panic.
+    pub fn new_with_current_thread(current_thread: &Arc<Thread>) -> Self {
+        let waiter = current_thread.waker().try_create_waiter().unwrap();
+        Self {
+            current_thread: current_thread.clone(),
+            waiter,
+            sig_mask: SigMask::new_empty(),
+        }
+    }
+
+    /// Sets the signal mask for this `Pauser`.
+    pub fn with_mask(mut self, sig_mask: SigMask) -> Self {
+        self.sig_mask = sig_mask;
+        self
+    }
+
+    /// Pauses the execution of current thread until the `cond` is met ( i.e., `cond()`
+    /// returns `Some(_)` ), or some signal is received by current thread or process.
+    ///
+    /// # Errors
+    ///
+    /// If some signal is received before `cond` is met, this method will returns `Err(EINTR)`.
+    pub fn pause_until<F, R>(&self, cond: F) -> Result<R>
+    where
+        F: FnMut() -> Option<R>,
+    {
+        self.do_pause(cond, None)
+    }
+
+    /// Pauses the execution of current thread until the `cond` is met ( i.e., `cond()` returns
+    /// `Some(_)` ), or some signal is received by current thread or process, or the given
+    /// `timeout` is expired.
+    ///
+    /// # Errors
+    ///
+    /// If `timeout` is expired before the `cond` is met or some signal is received,
+    /// it will returns [`ETIME`].
+    ///
+    /// [`ETIME`]: crate::error::Errno::ETIME
+    pub fn pause_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Result<R>
+    where
+        F: FnMut() -> Option<R>,
+    {
+        self.do_pause(cond, Some(timeout))
+    }
+
+    fn do_pause<F, R>(&self, mut cond: F, timeout: Option<&Duration>) -> Result<R>
+    where
+        F: FnMut() -> Option<R>,
+    {
+        let posix_thread = self.current_thread.as_posix_thread();
+        let cond = || {
+            if let Some(res) = cond() {
+                return Some(Ok(res));
+            }
+
+            if posix_thread.is_some_and(|posix_thread| posix_thread.has_pending()) {
```

Comment:
Something bothers me is the timeout logics. we have `wait_until_or_timeout` which does not accept `cancel_cond`, so I may need to add the `cancel_cond` parameter to `wait_until_or_timeout`. In addition, the reason that I did not introduce `wait_until_or_signalled` is also because it will intersect with `wait_until_or_timeout`, and we may need to add `timeout: Option<Duration>` parameter to `wait_until_or_signalled` or implement additional API like `wait_until_or_signalled_or_timeout`.

I am thinking if we need something like a `WaitConfig`, which contains cond and cancel_cond. Then we could add additional cond and cancel_cond through APIs like set_timeout and set_signalled. What do you think?

Reply:
> we may need to add `timeout: Option<Duration>` parameter to `wait_until_or_signalled`

Actually, this seems acceptable. We can continue to use "pause" to represent `wait_..._or_signalled` and just call the method as `pause_until_or_timeout`.